src/libexpr/eval.cc

```diff
@@ -711,128 +713,354 @@ std::optional<EvalState::Doc> EvalState::getDoc(Value & v)
+
+    throw error;
 }
 
 void EvalState::throwEvalError(const char * s, const std::string & s2) const
 {
-    throw EvalError(s, s2);
+    auto error = EvalError(s, s2);
+
+    debugLastTrace(error);
+
+    throw error;
 }
 
 void EvalState::throwEvalError(const PosIdx pos, const Suggestions & suggestions, const char * s,
-    const std::string & s2) const
+    const std::string & s2, Env & env, Expr & expr) const
 {
-    throw EvalError(ErrorInfo {
+    auto error = EvalError(ErrorInfo{
         .msg = hintfmt(s, s2),
         .errPos = positions[pos],
         .suggestions = suggestions,
     });
+
+    if (debuggerHook)
+        debuggerHook(&error, env, expr);
+
+    throw error;
 }
 
 void EvalState::throwEvalError(const PosIdx pos, const char * s, const std::string & s2) const
 {
-    throw EvalError(ErrorInfo {
+    auto error = EvalError({
         .msg = hintfmt(s, s2),
         .errPos = positions[pos]
     });
+
+    debugLastTrace(error);
+
+    throw error;
+}
+
+void EvalState::throwEvalError(const PosIdx pos, const char * s, const std::string & s2, Env & env, Expr & expr) const
+{
+    auto error = EvalError({
+        .msg = hintfmt(s, s2),
+        .errPos = positions[pos]
+    });
+
+    if (debuggerHook)
+        debuggerHook(&error, env, expr);
+
+    throw error;
 }
 
-void EvalState::throwEvalError(const char * s, const std::string & s2, const std::string & s3) const
+void EvalState::throwEvalError(const char * s, const std::string & s2,
+    const std::string & s3) const
 {
-    throw EvalError(s, s2, s3);
+    auto error = EvalError({
+        .msg = hintfmt(s, s2),
+        .errPos = positions[noPos]
+    });
+
+    debugLastTrace(error);
+
+    throw error;
 }
 
 void EvalState::throwEvalError(const PosIdx pos, const char * s, const std::string & s2,
     const std::string & s3) const
 {
-    throw EvalError({
-        .msg = hintfmt(s, s2, s3),
+    auto error = EvalError({
+        .msg = hintfmt(s, s2),
         .errPos = positions[pos]
     });
+
+    debugLastTrace(error);
+
+    throw error;
 }
 
-void EvalState::throwEvalError(const PosIdx p1, const char * s, const Symbol sym, const PosIdx p2) const
+void EvalState::throwEvalError(const PosIdx pos, const char * s, const std::string & s2,
+    const std::string & s3, Env & env, Expr & expr) const
+{
+    auto error = EvalError({
+        .msg = hintfmt(s, s2),
+        .errPos = positions[pos]
+    });
+
+    if (debuggerHook)
+        debuggerHook(&error, env, expr);
+
+    throw error;
+}
+
+void EvalState::throwEvalError(const PosIdx p1, const char * s, const Symbol sym, const PosIdx p2, Env & env, Expr & expr) const
 {
     // p1 is where the error occurred; p2 is a position mentioned in the message.
-    throw EvalError({
+    auto error = EvalError({
         .msg = hintfmt(s, symbols[sym], positions[p2]),
         .errPos = positions[p1]
     });
+
+    if (debuggerHook)
+        debuggerHook(&error, env, expr);
+
+    throw error;
```

Comment:
There is a lot of boilerplate like this, which could be avoided with a helper function that optionally starts the debugger and then throws the exception, e.g.
```
    throwError(EvalError(...));
```

Reply:
Ok I [tried that](https://github.com/bburdette/nix/commit/2c9fafdc9e43f6da39c289888dedbbbf0ea0b208), and remembered why I didn't do it before!  The exception that gets thrown ends up being a generic Error rather than a specific subclass, which breaks `tryEval` and probably other code that depends on the exception type.  

But I did eliminate some boilerplate with a `debugError` function.

```
void debugError(Error * e, Env & env, Expr & expr)
{
    if (debuggerHook)
        debuggerHook(e, env, expr);
}
```

Reply:
the template function approached worked out.  so now these:
```
    debugLastTrace(error);

    throw error;
```

look like this:
```
    debugThrowLastTrace(error);
```